builder/src/lib.rs

```diff
@@ -1,10 +1,247 @@
+#[proc_macro_derive(Builder, attributes(builder))]
 pub fn derive(input: TokenStream) -> TokenStream {
-    let _ = input;
+    let input = parse_macro_input!(input as DeriveInput);
+    let name = &input.ident;
+    let buildername = Ident::new(&format!("{}Builder", name), input.ident.span().clone());
 
-    unimplemented!()
+    let (fields, fields_init, setters, build) = match input.data {
+        Data::Struct(ref data) => {
+            match data.fields {
+                Fields::Named(ref fields) => {
+                    let builders = match get_fields_builders(&fields) {
+                        Some(b) => b,
+                        None => return TokenStream::from(quote! {
+                            compile_error!("missing or invalid `each` parameter in builder");
+                        }),
+                    };
+                    let optional_fields = get_optional_fields(&fields);
+                    (
+                        expand_field_definitions(&fields, &optional_fields, &builders),
+                        expand_field_initializers(&fields, &builders),
+                        expand_field_setters(&fields, &optional_fields, &builders),
+                        expand_build(&name, &fields, &optional_fields, &builders),
+                    )
+                }
+                _ => { (quote!{}, quote!{}, quote!{}, quote!{}) },
+            }
+        },
+        _ => unimplemented!(),
+    };
+
+    let expanded = quote! {
+        impl #name {
+            pub fn builder() -> #buildername {
+                #buildername { #fields_init }
+            }
+        }
+
+        pub struct #buildername {
+            #fields
+        }
+
+        impl #buildername {
+            #setters
+
+            pub fn build(&mut self) -> std::result::Result<#name, std::boxed::Box<dyn std::error::Error>> {
+                #build
+            }
+        }
+    };
+    TokenStream::from(expanded)
+}
+
+fn expand_field_definitions(fields: &syn::FieldsNamed, optional_fields: &HashSet<String>, builders: &HashMap<String, String>) -> TokenStream2 {
+    let f = fields.named.iter().map(|f| {
+        let ident = &f.ident;
+        let ty = &f.ty;
+        let identstr = f.ident.as_ref().map(|x| format!("{}", x)).unwrap_or("".to_owned());
+        if optional_fields.contains(&identstr) || builders.contains_key(&identstr) {
+            quote! {
+                #ident: #ty
+            }
+        } else {
+            quote! {
+                #ident: std::option::Option<#ty>
+            }
+        }
+    });
+    quote! { #(#f,)* }
+}
+
+fn expand_field_initializers(fields: &syn::FieldsNamed, builders: &HashMap<String, String>) -> TokenStream2 {
+    let f1 = fields.named.iter().filter(|f| {
+        !builders.contains_key(&f.ident.as_ref().unwrap().to_string())
+    }).map(|f| {
+        let ident = &f.ident;
+        quote! {
+            #ident: std::option::Option::None
+        }
+    });
+    let f2 = builders.keys().map(|k| {
```

Comment:
I try to avoid having procedural macro code in which the generated code depends on the random iteration order of a HashMap. I find that understanding the generated code tends to be easier if it is deterministic.

It is fine to use HashMaps internally as long as all code that gets generated is independent of HashMap iteration order.